lib/friendly_shipping/services/ups_json/generate_labels_payload.rb

```diff
@@ -0,0 +1,207 @@
+          end
+
+          if options.negotiated_rates
+            payload[:ShipmentRequest][:Shipment][:ShipmentRatingOptions] = { NegotiatedRatesIndicator: "X" }
+          end
+
+          payload[:ShipmentRequest][:Shipment][:ShipmentServiceOptions] = {
+            UPScarbonneutralIndicator: options.carbon_neutral ? "X" : nil,
+            SaturdayDeliveryIndicator: options.saturday_delivery ? "X" : nil,
+            LabelDelivery: {
+              LabelLinksIndicator: "X"
+            }
+          }.compact
+
+          if international?(shipment) && options.terms_of_shipment_code == "DDP" && options.billing_options.bill_third_party
+            payload[:ShipmentRequest][:Shipment][:PaymentInformation][:ShipmentCharge].append(build_ddp_billing_info(options))
+          end
+
+          if international?(shipment) && options.paperless_invoice
+            payload[:ShipmentRequest][:Shipment][:ShipmentServiceOptions][:InternationalForms] = international_forms(shipment, options)
+          end
+
+          if international?(shipment)
+            unless options.return_service_code
+              international_forms_additions = {}
+              sold_to_location = options.sold_to || shipment.destination
+              international_forms_additions[:Contacts] = { SoldTo: GenerateAddressHash.call(location: sold_to_location, international: true) }
+              international_forms_additions[:Contacts][:SoldTo].merge(
+                {
+                  Phone: { Number: sold_to_location.phone },
+                  TaxIdentificationNumber: sold_to_location.try(:tax_id_number),
+                  EmailAddress: sold_to_location.email
+                }
+              ).compact
+              payload[:ShipmentRequest][:Shipment][:ShipmentServiceOptions][:InternationalForms].merge!(international_forms_additions)
+            end
+
+            if shipment.origin.country.code == 'US' && ['CA', 'PR'].include?(shipment.destination.country.code)
+              # Required for shipments from the US to Puerto Rico or Canada
+              # We'll assume USD as the origin country is the United States here.
+              total_value = shipment.packages.inject(Money.new(0, 'USD')) do |shipment_sum, package|
+                shipment_sum + package.items.inject(Money.new(0, 'USD')) do |package_sum, item|
+                  package_sum + (item.cost || Money.new(0, 'USD'))
+                end
+              end
+              payload[:ShipmentRequest][:Shipment][:InvoiceLineTotal] = {
+                CurrencyCode: 'USD',
+                MonetaryValue: total_value.to_f.to_s
+              }
+            end
+
+            contents_description = shipment.packages.flat_map do |package|
+              package.items.map(&:description)
+            end.compact.uniq.join(', ').slice(0, 50)
+            payload[:ShipmentRequest][:Shipment][:Description] = contents_description
```

Comment:
Didn't you already handle this above on lines 11-13 and 23?

Reply:
Yes, nice catch.